bot/exts/utilities/bookmark.py

```diff
@@ -16,14 +18,51 @@
 # Number of seconds to wait for other users to bookmark the same message
 TIMEOUT = 120
 BOOKMARK_EMOJI = "📌"
+MESSAGE_NOT_FOUND_ERROR = (
+    "You must either provide a valid message to bookmark, or reply to one."
+    "\n\nThe lookup strategy for a message is as follows (in order):"
+    "\n1. Lookup by '{channel ID}-{message ID}' (retrieved by shift-clicking on 'Copy ID')"
+    "\n2. Lookup by message ID (the message **must** have been sent after the bot last started)"
+    "\n3. Lookup by message URL"
+)
+
+
+class AlreadyBookmarkedError(commands.UserInputError):
+    """Raised when a user tries to bookmark a message they have already bookmarked."""
 
 
 class Bookmark(commands.Cog):
     """Creates personal bookmarks by relaying a message link to the user's DMs."""
 
+    # A lookup of what messages each member has bookmarked.
+    # Used to stop members from bookmarking the same message twice.
+    # {member id: serialised json list of message ids}
+    all_member_bookmarked_messages = RedisCache()
+
     def __init__(self, bot: Bot):
         self.bot = bot
 
+    async def get_member_bookmarked_messages(self, member: discord.Member) -> list[int]:
+        """De-serialise and return a list of a member's bookmarked messages."""
+        single_members_bookmarked_messages = await self.all_member_bookmarked_messages.get(member.id, "[]")
+        return json.loads(single_members_bookmarked_messages)
+
+    async def update_member_bookmarked_messages(
+        self,
+        member: discord.Member,
+        message_id: int,
+        add_or_remove: Literal["add", "remove"]
+    ) -> None:
+        """De-serialise, run specified action, serialise, and store the messages a member has bookmarked."""
+        single_members_bookmarked_messages = await self.get_member_bookmarked_messages(member)
+        if message_id not in single_members_bookmarked_messages:
+            # Member deleted a DM message other than a bookmark
+            return
+
+        action = list.append if add_or_remove == "add" else list.remove
+        action(single_members_bookmarked_messages, message_id)
```

Comment:
Since the previous if-statement should be removed, this can now raise a ValueError.

Reply:
For `list.remove`? The if statement can be kept for `add_or_remove == "remove"` (even though it may be guarded by [this if `.unbm` is only for actual bookmarks](https://github.com/python-discord/sir-lancebot/pull/929#discussion_r743390980)).

Reply:
Rather than use strings at all, what about an enum of `Action.DELETE` and `Action.ADD` or something similar?